File: packages/python/tryaii_dre/scoring/engine.py
"""
Dynamic model scoring engine.

Combines benchmark performance, cost, and speed into a single score
weighted by user priorities. This is the heart of the routing logic.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from tryaii_dre.registry.models import ModelInfo
from tryaii_dre.scoring.benchmarks import BenchmarkNormalizer
from tryaii_dre.scoring.priorities import DEFAULT_PRIORITIES, Priorities


@dataclass
class ModelScore:
    """Detailed score breakdown for a single model."""

    model_id: str
    final_score: float  # 0-1 combined score
    quality_score: float  # 0-1 benchmark quality
    cost_score: float  # 0-1 (higher = cheaper)
    speed_score: float  # 0-1 (higher = faster)
    quality_contribution: float
    cost_contribution: float
    speed_contribution: float
    top_benchmarks: list[tuple[str, float]]  # Most relevant benchmarks for this model
    reasoning: str  # Human-readable explanation


# Speed tier -> numeric score
SPEED_SCORES: dict[str, float] = {
    "very fast": 1.0,
    "fast": 0.8,
    "medium": 0.6,
    "slow": 0.3,
    "very slow": 0.1,
}


class ScoringEngine:
    """
    Scores models against a classified prompt.

    Takes benchmark similarity scores (from the classifier) and user priorities,
    then ranks all available models using a three-factor weighted algorithm:

        final = (quality * qW + cost * cW + speed * sW) / (qW + cW + sW)

    Where weights are derived from user priorities (1-5 scale).
    """

    def __init__(self, normalizer: Optional[BenchmarkNormalizer] = None):
        self._normalizer = normalizer or BenchmarkNormalizer()

    def score_models(
        self,
        models: list[ModelInfo],
        benchmark_similarities: dict[str, float],
        priorities: Priorities = DEFAULT_PRIORITIES,
        top_k: int = 5,
    ) -> list[ModelScore]:
        """
        Score and rank models based on benchmark similarities and priorities.

        Args:
            models: Available models to score.
            benchmark_similarities: Cosine similarity of user prompt to each benchmark
                                    centroid. Keys are benchmark names, values are 0-1.
            priorities: User priority weights.
            top_k: Return top K models.

        Returns:
            Sorted list of ModelScore objects (highest score first).
        """
        # Use top 3 most relevant benchmarks for scoring
        top_benchmarks = sorted(
            benchmark_similarities.items(), key=lambda x: x[1], reverse=True
        )[:3]
        top_benchmark_dict = dict(top_benchmarks)

        scores: list[ModelScore] = []

        for model in models:
            score = self._score_single_model(model, top_benchmark_dict, priorities)
            if score is not None:
                scores.append(score)

        # Sort by final score descending
        scores.sort(key=lambda s: s.final_score, reverse=True)

        # Normalize to 0.1-0.95 range (best model ~ 0.95)
        if scores:
            max_raw = scores[0].final_score
            min_raw = scores[-1].final_score if len(scores) > 1 else 0

            for s in scores:
                if max_raw == min_raw:
                    s.final_score = 0.5
                else:
                    normalized = (s.final_score - min_raw) / (max_raw - min_raw)
                    s.final_score = round(0.1 + 0.85 * normalized, 4)

        return scores[:top_k]

    def _score_single_model(
        self,
        model: ModelInfo,
        top_benchmarks: dict[str, float],
        priorities: Priorities,
    ) -> Optional[ModelScore]:
        """Score a single model against the benchmark similarities."""

        # --- Quality score ---
        weighted_quality_sum = 0.0
        total_similarity_weight = 0.0
        model_top_benchmarks: list[tuple[str, float]] = []

        for benchmark_name, user_similarity in top_benchmarks.items():
            model_bench_score = model.benchmark_scores.get(benchmark_name)
            if model_bench_score is None:
                continue

            normalized = self._normalizer.normalize(benchmark_name, model_bench_score)
            weighted_quality_sum += user_similarity * normalized
            total_similarity_weight += user_similarity
            model_top_benchmarks.append((benchmark_name, normalized))

        if total_similarity_weight == 0:
            return None

        quality_score = weighted_quality_sum / total_similarity_weight

        # --- Cost score ---
        cost_score = 0.0
        if model.pricing:
            avg_cost = (model.pricing.input_per_1k + model.pricing.output_per_1k) / 2
            # Normalize against $0.10/1k tokens baseline
            cost_score = max(0.0, 1.0 - (avg_cost / 0.1))

        # --- Speed score ---
        speed_score = 0.0
        if model.latency:
            speed_score = SPEED_SCORES.get(model.latency, 0.3)

        # --- Combine with priority weights ---
        q_weight = priorities.quality_weight
        c_weight = priorities.cost_weight
        s_weight = priorities.speed_weight

        q_contrib = quality_score * q_weight
        c_contrib = cost_score * c_weight
        s_contrib = speed_score * s_weight

        total_weight = q_weight + c_weight + s_weight
        final = (q_contrib + c_contrib + s_contrib) / total_weight
        final = max(0.0, min(1.0, final))

        # Generate reasoning
        top_bench_str = ", ".join(
            f"{b} ({s:.0%})" for b, s in model_top_benchmarks[:2]
        )
        reasoning = f"Quality: {quality_score:.2f} on [{top_bench_str}]"
        if cost_score > 0:
            reasoning += f" | Cost efficiency: {cost_score:.2f}"
        if speed_score > 0:
            reasoning += f" | Speed: {speed_score:.2f} ({model.latency})"

        return ModelScore(
            model_id=model.model_id,
            final_score=final,
            quality_score=round(quality_score, 4),
            cost_score=round(cost_score, 4),
            speed_score=round(speed_score, 4),
            quality_contribution=round(q_contrib, 4),
            cost_contribution=round(c_contrib, 4),
            speed_contribution=round(s_contrib, 4),
            top_benchmarks=model_top_benchmarks,
            reasoning=reasoning,
        )

File: packages/node/src/scoring/engine.ts
/**
 * Dynamic model scoring engine.
 *
 * Combines benchmark performance, cost, and speed into a single score
 * weighted by user priorities. This is the heart of the routing logic.
 */

import { ModelInfo } from '../registry/models.js';
import { BenchmarkNormalizer } from './benchmarks.js';
import { DEFAULT_PRIORITIES, Priorities } from './priorities.js';

export interface ModelScore {
  modelId: string;
  finalScore: number;       // 0-1 combined score
  qualityScore: number;     // 0-1 benchmark quality
  costScore: number;        // 0-1 (higher = cheaper)
  speedScore: number;       // 0-1 (higher = faster)
  qualityContribution: number;
  costContribution: number;
  speedContribution: number;
  topBenchmarks: Array<[string, number]>; // Most relevant benchmarks for this model
  reasoning: string;        // Human-readable explanation
}

/** Speed tier -> numeric score. */
export const SPEED_SCORES: Record<string, number> = {
  'very fast': 0.5,
  'fast': 0.4,
  'medium': 0.3,
  'slow': 0.2,
  'very slow': 0.1,
};

/**
 * Scores models against a classified prompt.
 *
 * Takes benchmark similarity scores (from the classifier) and user priorities,
 * then ranks all available models using a three-factor weighted algorithm:
 *
 *     final = (quality * qW + cost * cW + speed * sW) / (qW + cW + sW)
 *
 * Where weights are derived from user priorities (1-5 scale).
 */
export class ScoringEngine {
  private _normalizer: BenchmarkNormalizer;

  constructor(normalizer?: BenchmarkNormalizer) {
    this._normalizer = normalizer ?? new BenchmarkNormalizer();
  }

  /**
   * Score and rank models based on benchmark similarities and priorities.
   */
  scoreModels(
    models: ModelInfo[],
    benchmarkSimilarities: Record<string, number>,
    priorities: Priorities = DEFAULT_PRIORITIES,
    topK = 5,
  ): ModelScore[] {
    // Use top 3 most relevant benchmarks for scoring
    const sortedBenchmarks = Object.entries(benchmarkSimilarities)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 3);

    const topBenchmarkDict: Record<string, number> = {};
    for (const [name, score] of sortedBenchmarks) {
      topBenchmarkDict[name] = score;
    }

    const scores: ModelScore[] = [];

    for (const model of models) {
      const score = this._scoreSingleModel(model, topBenchmarkDict, priorities);
      if (score !== null) {
        scores.push(score);
      }
    }

    // Sort by final score descending
    scores.sort((a, b) => b.finalScore - a.finalScore);

    // Normalize to 0.1-0.95 range (best model ~ 0.95)
    if (scores.length > 0) {
      const maxRaw = scores[0].finalScore;
      const minRaw = scores.length > 1 ? scores[scores.length - 1].finalScore : 0;

      for (const s of scores) {
        if (maxRaw === minRaw) {
          s.finalScore = 0.5;
        } else {
          const normalized = (s.finalScore - minRaw) / (maxRaw - minRaw);
          s.finalScore = Math.round((0.1 + 0.85 * normalized) * 10000) / 10000;
        }
      }
    }

    return scores.slice(0, topK);
  }

  private _scoreSingleModel(
    model: ModelInfo,
    topBenchmarks: Record<string, number>,
    priorities: Priorities,
  ): ModelScore | null {
    // --- Quality score ---
    let weightedQualitySum = 0;
    let totalSimilarityWeight = 0;
    const modelTopBenchmarks: Array<[string, number]> = [];

    for (const [benchmarkName, userSimilarity] of Object.entries(topBenchmarks)) {
      const modelBenchScore = model.benchmarkScores[benchmarkName];
      if (modelBenchScore == null) continue;

      const normalized = this._normalizer.normalize(benchmarkName, modelBenchScore);
      weightedQualitySum += userSimilarity * normalized;
      totalSimilarityWeight += userSimilarity;
      modelTopBenchmarks.push([benchmarkName, normalized]);
    }

    if (totalSimilarityWeight === 0) return null;

    const qualityScore = weightedQualitySum / totalSimilarityWeight;

    // --- Cost score ---
    let costScore = 0;
    if (priorities.cost > 1 && model.pricing) {
      const avgCost = (model.pricing.inputPer1k + model.pricing.outputPer1k) / 2;
      // Normalize against $0.10/1k tokens baseline
      costScore = Math.max(0.0, 1.0 - avgCost / 0.1);
    }

    // --- Speed score ---
    let speedScore = 0;
    if (priorities.speed > 1 && model.latency) {
      speedScore = SPEED_SCORES[model.latency] ?? 0.3;
    }

    // --- Combine with priority weights ---
    const qWeight = priorities.qualityWeight;
    const cWeight = priorities.costWeight;
    const sWeight = priorities.speedWeight;

    const qContrib = qualityScore * qWeight;
    const cContrib = costScore * cWeight;
    const sContrib = speedScore * sWeight;

    const totalWeight = qWeight + cWeight + sWeight;
    let final = (qContrib + cContrib + sContrib) / totalWeight;
    final = Math.max(0.0, Math.min(1.0, final));

    // Generate reasoning
    const topBenchStr = modelTopBenchmarks
      .slice(0, 2)
      .map(([b, s]) => `${b} (${Math.round(s * 100)}%)`)
      .join(', ');

    let reasoning = `Quality: ${qualityScore.toFixed(2)} on [${topBenchStr}]`;
    if (costScore > 0) {
      reasoning += ` | Cost efficiency: ${costScore.toFixed(2)}`;
    }
    if (speedScore > 0) {
      reasoning += ` | Speed: ${speedScore.toFixed(2)} (${model.latency})`;
    }

    return {
      modelId: model.modelId,
      finalScore: final,
      qualityScore: Math.round(qualityScore * 10000) / 10000,
      costScore: Math.round(costScore * 10000) / 10000,
      speedScore: Math.round(speedScore * 10000) / 10000,
      qualityContribution: Math.round(qContrib * 10000) / 10000,
      costContribution: Math.round(cContrib * 10000) / 10000,
      speedContribution: Math.round(sContrib * 10000) / 10000,
      topBenchmarks: modelTopBenchmarks,
      reasoning,
    };
  }
}
